design/external-references.md

```diff
@@ -0,0 +1,394 @@
+* The `weak` option is only relevant for synth-time resolution. If it is `false` (which is the default), the CDK will automatically embed a `Metadata` entry on the consuming resource with an `Fn::ImportValue`. This will force CloudFormation to take a strong reference on the export, even through the actual value is concretely resolved during synthesis. This ensures, for example, that the producing stack can't be deleted as long as there stacks consuming the exported values. This behavior (which is the default), can be disabled by settings `weak: true`, in which case the `Fn::ImportValue` will simply not be included.
+
+On top of these two methods, we can now define import and export methods for serializable objects:
+
+```ts
+class Stack {
+  exportObject(exportName: string, obj: ISerializable): void;
+  importObject(exportName: string, options?: ImportOptions): IDeserializationContext;
+}
+```
+
+The `exportObject` method will invoke `obj.serialize` with a serialization context "bound" to this export name. This means the export name will be used as a prefix to all written keys. `writeObject` will be implemented with a nested serialization context that adds another component to the export name prefix.
+
+The `importObject` method will be used like this:
+
+```ts
+const importedBucket = Bucket.deserializeBucket(stack.importObject('MyBucketExportName'));
+```
+
+The method will return a deserialization context that's bound to the export name. Similarly to the serialization context, it will prefix all values read through `readString` with the export name, and so forth with `readObject`.
+
+As mentioned about, the default resolve type for imports is `Synthesis` (with strong-references). This means that the values returned by `readString` will be actual concrete values. If users opt-in to deploy-time resolution (by setting using `Deployment` resolve type), the values returned will be tokens. In some cases this would be fine, but there could be constructs that cannot deal with opaque values (i.e. if the value is an list of strings and needs to be deconstructed). In those cases (REQ8), constructs should invoke `readString` with `allowUnresolved: false` to indicate that this specific value cannot be a token.
+
+#### Export names
+
+AWS CloudFormation export names must be unique within an environment (account/region), and they will be formed by concatenating root `exportName` and all the keys that lead to a value in the serialization tree.
+
+We will use `-` as a component separator devising fully qualified export names. To avoid collisions, if the main export name or any subsequent serialization key includes a `-` it will be removed.
+
+Since AWS CloudFormation has a limit on export name length, and we wouldn't want to restrict the serialization depth, the import/export serializer should trim the name and add a hash of the full name, but only if the total length exceeds the limit.
```

Comment:
Length is only one of the limits. There's also illegal characters. Export names must also be account+region-unique, so the stack name should be prepended somehow, too.